nameresolution/consul/watcher.go

```diff
@@ -0,0 +1,342 @@
+package consul
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+)
+
+const (
+	// retryInterval is the base retry value.
+	retryInterval = 5 * time.Second
+
+	// maximum back off time, this is to prevent exponential runaway.
+	maxBackoffTime = 180 * time.Second
+)
+
+type watchPlan struct {
+	expired               bool
+	lastParamVal          blockingParamVal
+	lastResult            map[serviceIdentifier]bool
+	options               *consul.QueryOptions
+	configuredQueryFilter string
+	failures              int
+}
+
+type blockingParamVal interface {
+	equal(other blockingParamVal) bool
+	next(previous blockingParamVal) blockingParamVal
+}
+
+type waitIndexVal uint64
+
+// Equal implements BlockingParamVal.
+func (idx waitIndexVal) equal(other blockingParamVal) bool {
+	if otherIdx, ok := other.(waitIndexVal); ok {
+		return idx == otherIdx
+	}
+
+	return false
+}
+
+// Next implements BlockingParamVal.
+func (idx waitIndexVal) next(previous blockingParamVal) blockingParamVal {
+	if previous == nil {
+		return idx
+	}
+	prevIdx, ok := previous.(waitIndexVal)
+	if ok && prevIdx == idx {
+		// This value is the same as the previous index, reset
+		return waitIndexVal(0)
+	}
+
+	return idx
+}
+
+type serviceIdentifier struct {
+	serviceName string
+	serviceID   string
+	node        string
+}
+
+func getServiceIdentifier(c *consul.HealthCheck) serviceIdentifier {
+	return serviceIdentifier{
+		serviceID:   c.ServiceID,
+		serviceName: c.ServiceName,
+		node:        c.Node}
+}
+
+func getHealthByService(checks consul.HealthChecks) map[serviceIdentifier]bool {
+	healthByService := make(map[serviceIdentifier]bool)
+	for _, check := range checks {
+		id := getServiceIdentifier(check)
+
+		if state, ok := healthByService[id]; !ok {
+			// Init to healthy
+			healthByService[id] = true
+		} else if !state {
+			continue
+		}
+
+		if check.Status != consul.HealthPassing {
+			healthByService[id] = false
+		}
+	}
+
+	return healthByService
+}
+
+func (p *watchPlan) getChangedServices(newResult map[serviceIdentifier]bool) map[string]struct{} {
+	changedServices := make(map[string]struct{})
+
+	// get changed services
+	for newKey, newValue := range newResult {
+		if oldValue, ok := p.lastResult[newKey]; ok && newValue == oldValue {
+			continue
+		}
+
+		changedServices[newKey.serviceName] = struct{}{}
+	}
+
+	for oldKey := range p.lastResult {
+		if _, ok := newResult[oldKey]; !ok {
+			changedServices[oldKey.serviceName] = struct{}{}
+		}
+	}
+
+	return changedServices
+}
+
+func (p *watchPlan) buildServiceNameFilter(services []string) {
+	var nameFilters = make([]string, len(services))
+
+	for i, v := range services {
+		nameFilters[i] = fmt.Sprintf("ServiceName==\"%s\"", v)
```

Comment:
fmt.Sprintf performs multiple allocations on the heap, so for simple string concatenation (esp. in a hot path), best to use `+`
```suggestion
		nameFilters[i] = `ServiceName=="`+v+`"`
```